mava/systems/sac/ff_isac.py

```diff
@@ -0,0 +1,565 @@
+        params, opt_states, act_loss_info = lax.cond(
+            t % cfg.system.policy_frequency == 0,  # TD 3 Delayed update support
+            update_actor_and_alpha,
+            # just return same params and opt_states and 0 for losses
+            lambda params, opt_states, *_: (
+                params,
+                opt_states,
+                {"actor_loss": 0.0, "alpha_loss": 0.0},
+            ),
+            params,
+            opt_states,
+            data,
+            actor_key,
+        )
+
+        losses = q_loss_info | act_loss_info
+
+        return (buffer_state, params, opt_states, t, key), losses
+
+    def act(
+        carry: Tuple[FrozenVariableDict, Array, State, BufferState, chex.PRNGKey], _: Any
+    ) -> Tuple[Tuple[FrozenVariableDict, Array, State, BufferState, chex.PRNGKey], Dict]:
+        """Acting loop: select action, step env, add to buffer."""
+        actor_params, obs, env_state, buffer_state, key = carry
+
+        pi = actor.apply(actor_params, obs)
+        action = pi.sample(seed=key)
+
+        next_obs, env_state, buffer_state, metrics = step(action, obs, env_state, buffer_state)
+        return (actor_params, next_obs, env_state, buffer_state, key), metrics
+
+    def explore(carry: LearnerState, _: Any) -> Tuple[LearnerState, Metrics]:
+        """Take random actions to fill up buffer at the start of training."""
+        obs, env_state, buffer_state, _, _, t, key = carry
+        key, explore_key = jax.random.split(key)
+        action = jax.random.uniform(explore_key, full_action_shape)
+        next_obs, env_state, buffer_state, metrics = step(action, obs, env_state, buffer_state)
+
+        t += cfg.system.n_envs
+        learner_state = carry._replace(
+            obs=next_obs, env_state=env_state, buffer_state=buffer_state, t=t, key=key
+        )
+        return learner_state, metrics
+
+    scanned_update = lambda state: lax.scan(update_epoch, state, None, length=cfg.system.epochs)
+    scanned_act = lambda state: lax.scan(act, state, None, length=cfg.system.rollout_length)
+
+    # Act loop -> sample -> update loop
+    def update_step(carry: LearnerState, _: Any) -> Tuple[LearnerState, Tuple[Metrics, Metrics]]:
+        """Act, sample, learn."""
+
+        obs, env_state, buffer_state, params, opt_states, t, key = carry
+        key, act_key, learn_key = jax.random.split(key, 3)
+        # Act
+        act_state = (params.actor, obs, env_state, buffer_state, act_key)
+        (_, next_obs, env_state, buffer_state, _), metrics = scanned_act(act_state)
+
+        # Sample and learn
+        learn_state = (buffer_state, params, opt_states, t, learn_key)
+        (buffer_state, params, opt_states, _, _), losses = scanned_update(learn_state)
+
+        t += cfg.system.n_envs * cfg.system.rollout_length
+        return (
+            LearnerState(next_obs, env_state, buffer_state, params, opt_states, t, key),
+            (metrics, losses),
+        )
```

Comment:
I like how visible the overall structure here is - the name might suggest, though, that this updates the "step" function defined earlier. The docstrings go a long way in providing clarification.

Reply:
Good point, not sure what else to call it though, it's like the main algorithm step :thinking:

Reply:
I'm happy with `update_step`, we could call it like `main_step` or something but I think `update_step` works ok here